sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/BagUserState.java

```diff
@@ -130,13 +132,31 @@ public void asyncClose() throws Exception {
     if (!newValues.isEmpty()) {
       ByteStringOutputStream out = new ByteStringOutputStream();
       for (T newValue : newValues) {
-        // TODO: Replace with chunking output stream
+        int previousSize = out.size();
         valueCoder.encode(newValue, out);
+        if (out.size() > BAG_APPEND_BATCHING_LIMIT && previousSize > 0) {
+          // Respect the limit by outputting the previous batch of elements.
+          beamFnStateClient.handle(
+              request
+                  .toBuilder()
+                  .setAppend(
+                      StateAppendRequest.newBuilder()
+                          .setData(out.consumePrefixToByteString(previousSize))));
+        }
+        if (out.size() > BAG_APPEND_BATCHING_LIMIT) {
```

Comment:
This seems to imply 10MB is a soft limit. 

If so, I think it'd be easier to simply emit once the limit is exceeded, rather than introduce the (more complicated to reason about, though I think it's correct) consumePrefixToByteString just to avoid the one extra element.

Reply:
The 10MB is a soft-limit as that seems large enough to get efficiency wins of batching over rpcs while keeping things more granular for paging in later. But for the dataflow streaming backend there is a hard-limit of 100MB.  Incoming elements can be up to that limit, so if those elements are directly written to a bag I want to ensure we don't batch beyond that.

I will add some comments

Reply:
OK, that's the same soft limit we use on the (push) data channel. Let's avoid introducing the complexity of consumePrefixToByteString and just send + create a new object when we hit this threshold.

Reply:
Can you clarify? Here we don't have the encoded sizes until we actually encode, and for efficiency it is nice to encode with ByteStringOutputStream instead of creating a separate output stream for each value and concatenating to respect the limit.  Pulling off the prefix seemed like an ok solution to keep the performance of ByteStringOutputStream.  Were you thinking the performance hit of separate buffers was worth the simpler code?

Reply:
I was thinking of just having the loop

```
for (T newValue : newValues) {
  valueCoder.encode(newValue, out);
  if (out.size() > BAG_APPEND_BATCHING_LIMIT) {
    [send out.toByteStringAndReset()]
  }
}
...
```

rather than introducing the (I think correct, but complex to reason about) `consumePrefixToByteString`. True, we will fail on a element of size 100-x MB after writing x MB, for x < 10, but the question is whether that corner case is worth the additional complexity.

Reply:
Per offline discussion, keeping the existing behavior which ensures we don't increase the size of large elements through batching. This is important to respect runner limits such as in Dataflow. Input elements may be up to the limit and thus greedy batching could exceed the limit.